nullaway/src/main/java/com/uber/nullaway/handlers/AssertionHandler.java

```diff
@@ -0,0 +1,128 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.uber.nullaway.handlers;
+
+import static com.uber.nullaway.Nullness.NONNULL;
+
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.code.Types;
+import com.sun.tools.javac.util.Context;
+import com.sun.tools.javac.util.Name;
+import com.uber.nullaway.dataflow.AccessPath;
+import com.uber.nullaway.dataflow.AccessPathNullnessPropagation;
+import org.checkerframework.dataflow.cfg.node.MethodInvocationNode;
+import org.checkerframework.dataflow.cfg.node.Node;
+
+/** This Handler deals with assertions which ensure that their arguments cannot be null. */
+public class AssertionHandler extends BaseNoOpHandler {
+  public AssertionHandler() {
+    super();
+  }
+
+  @Override
+  public NullnessHint onDataflowVisitMethodInvocation(
+      MethodInvocationNode node,
+      Types types,
+      Context context,
+      AccessPathNullnessPropagation.SubNodeValues inputs,
+      AccessPathNullnessPropagation.Updates thenUpdates,
+      AccessPathNullnessPropagation.Updates elseUpdates,
+      AccessPathNullnessPropagation.Updates bothUpdates) {
+    Symbol.MethodSymbol callee = ASTHelpers.getSymbol(node.getTree());
+    if (callee == null) {
+      return NullnessHint.UNKNOWN;
+    }
+
+    if (!areMethodNamesInitialized()) {
+      initializeMethodNames(callee);
+    }
+
+    /**
+     * Look for statements of the form: assertThat(A).isNotNull()
+     *
+     * <p>A will not be NULL after this statement.
+     */
+    if (isMethodIsNotNull(callee)) {
+      Node receiver = node.getTarget().getReceiver();
+      if (receiver instanceof MethodInvocationNode) {
+        MethodInvocationNode receiver_method = (MethodInvocationNode) receiver;
+        Symbol.MethodSymbol receiver_symbol = ASTHelpers.getSymbol(receiver_method.getTree());
+        if (isMethodAssertThat(receiver_symbol)) {
+          Node arg = receiver_method.getArgument(0);
+          AccessPath ap = AccessPath.getAccessPathForNodeNoMapGet(arg);
+          if (ap != null) {
+            bothUpdates.set(ap, NONNULL);
+          }
+        }
+      }
+    }
+    return NullnessHint.UNKNOWN;
+  }
+
+  private boolean isMethodIsNotNull(Symbol.MethodSymbol methodSymbol) {
+    return matchesMethod(methodSymbol, isNotNull, isNotNullOwner);
+  }
+
+  private boolean isMethodAssertThat(Symbol.MethodSymbol methodSymbol) {
+    return matchesMethod(methodSymbol, assertThat, assertThatOwner);
+  }
+
+  private boolean matchesMethod(
+      Symbol.MethodSymbol methodSymbol, Name toMatchMethodName, Name toMatchOwnerName) {
+    return methodSymbol.name.equals(toMatchMethodName)
+        && methodSymbol.owner.getQualifiedName().equals(toMatchOwnerName);
+  }
+
+  private boolean areMethodNamesInitialized() {
+    return isNotNull != null;
```

Comment:
Actually, upon reflection, this whole thing is not thread safe, since this could execute concurrent with `initializeMethodNames` and return `true` when say `isNotNullOwner` and the rest are still uninitialized. So the other method being `synchronized` is mostly pointless.

My thoughts: 

* I remain unsure whether the same BugChecker instance, and therefore this class, will be called in parallel, we could just leave both methods unprotected and the then fix it if it turns out it can. (cc: @msridhar , do you know the answer here? )
* We could conservatively move this check to `initializeMethodNames` and just have that method return early if `isNotNull` is already set, then call `initializeMethodNames` unconditionally from `onDataflowVisitMethodInvocation`
* We could use `toString()` to match the methods, measure the overhead and leave it at that if it's not high. Early optimization being the root of all evil and all that :)

What do you think?

Reply:
Concurrency is only an issue with static fields, which we should avoid unless they are immutable values.  There should be a single BugChecker instance per javac instance and it should not be getting called from multiple threads.

Since the only static fields here are final Strings we should be fine.  But the `synchronized` on the private method is unnecessary.

Reply:
Thanks for the clarification. Removed `synchronized` on the method.